### PR TITLE
🚧Create branch validator service

### PR DESCRIPTION
**Pull request by CalvinWilkinson**
<h1 style="font-weight:bold" align="center">Preview Feature Pull Request</h1>
<h2 style="font-weight:bold" align="center">✅Code Review Checklist✅</h2>

- [x] The **_head(source)_** branch for this pull request is a **_preview feature_** branch, with the correct naming syntax, in the following example:
  - A **_head(source)_** branch is the branch being merged into a target branch.
    <details closed><summary>Example</summary>

      ``` xml
      Syntax: preview/feature/<issue-num>-<description>
      Example: preview/feature/123-my-preview-feature
      ```
    </details>

- [x] The **_head(source)_** branch for this pull request is created from a **_preview release_** branch, with the correct naming syntax, in the following example:
  <details closed><summary>Example</summary>

    ``` xml
    Syntax: preview/v<major>.<minor>.<patch>-preview.<prev-num>
    Example: preview/v1.2.3-preview.4
    ```
  </details>

- [x] The **_base(target)_** branch for this pull request is a **_preview release_** branch, with the correct naming syntax, in the following example:
  - A **_base(target)_** branch is the branch that the **_head(source)_** branch is merging into.

    <details closed><summary>Example</summary>

      ``` xml
      Syntax: preview/v<major>.<minor>.<patch>-preview.<prev-num>
      Example: preview/v1.2.3-preview.4
      ```
    </details>

💡For more information on branching, refer to the project's [branching documentation](../../Documentation/Branching.md).

- [x] Associated issue exists and is linked to this pull request.
  - One issue per pull request.

- [x] Pull request title matches the title of the associated issue.

- [x] A `preview` label has been added to the pull request.

- [x] The labels attached to this PR match the labels attached to the associated issue.

- [x] Project assigned matches the project of the associated issue.

- [x] Milestone assigned matches the milestone of the associated issue.

- [x] My code follows the coding style of this project.
  - The style is enforced by the **_.editorconfig_** files in the project and displayed as warnings.  If there is an edge case, reach out and let's discuss it.

- [x] I have manually tested my code changes to the best of my ability.

- [x] All tests passed locally.
  - This is required because unit tests might pass locally but not in the CI environment during the status check process or vice-versa.
  - Tests might pass on the developer's machine but not necessarily on the code reviewer's machine.
  - Status checks run for this pull request when merging into the **_preview/v\*.\*.\*-preview.\*_** branch.  These status checks run every time a change is pushed to the pull request.  These checks validate version syntax, tagging, builds, unit tests, and more.

### CODE REVIEW

File: CICD/Reactables/Core/ReactorUnsubscriber.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+// ReSharper disable InconsistentNaming
+using System;
+using System.Collections.Generic;
+using CICDSystem.Guards;
+
+namespace CICDSystem.Reactables.Core;
+
+/// <summary>
+/// A reactor unsubscriber for unsubscribing from an <see cref="Reactable{TData}"/>.
```
Comment (kselena) [inline CICD/Reactables/Core/ReactorUnsubscriber.cs:13]: grammar
```suggestion
/// A reactor unsubscriber for unsubscribing from a <see cref="Reactable{TData}"/>.
```

File: CICD/Reactables/Core/ReactorUnsubscriber.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        Reactor = reactor;
+    }
+
+    /// <summary>
+    /// Gets the reactors of this unsubscription.
+    /// </summary>
+    public IReactor<T> Reactor { get; }
+
+    /// <summary>
+    /// Gets the total number of current subscriptions that an <see cref="Reactable{TData}"/> has.
```
Comment (kselena) [inline CICD/Reactables/Core/ReactorUnsubscriber.cs:42]: grammar
```suggestion
    /// Gets the total number of current subscriptions that a <see cref="Reactable{TData}"/> has.
```

File: CICD/Reactables/Core/ReactorUnsubscriber.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+using System.Collections.Generic;
+using CICDSystem.Guards;
+
+namespace CICDSystem.Reactables.Core;
+
+/// <summary>
+/// A reactor unsubscriber for unsubscribing from an <see cref="Reactable{TData}"/>.
+/// </summary>
+/// <typeparam name="T">
+///     The type of data that is pushed to all of the subscribed <see cref="Reactor{T}"/>s.
```
Comment (kselena) [inline CICD/Reactables/Core/ReactorUnsubscriber.cs:16]:
```suggestion
/// The type of data that is pushed to all of the subscribed <see cref="Reactor{T}"/>s.
```

File: CICD/Services/BranchValidatorService.cs [hunk 4]
```diff
@@ -0,0 +1,700 @@
...
+
+        this.orValues.Add(listToAdd.ToList());
+
+        this.andValues.Clear();
+
+        return this;
+    }
+
+    /// <inheritdoc/>
+    /// <exception cref="ArgumentNullException">Thrown if the <paramref name="branch"/> is null or empty..</exception>
```
Comment (kselena) [inline CICD/Services/BranchValidatorService.cs:523]: Remove extra period
```suggestion
    /// <exception cref="ArgumentNullException">Thrown if the <paramref name="branch"/> is null or empty.</exception>
```

File: CICD/Services/BranchValidatorService.cs [hunk 5]
```diff
@@ -0,0 +1,700 @@
...
+        if (string.IsNullOrEmpty(branch))
+        {
+            throw new Exception("Could not detect the current repository branch.  Does a repository exist?");
+        }
+
+        return branch.IsCorrectBranch(branchPattern);
+    }
+
+    /// <inheritdoc/>
+    /// <exception cref="Exception">Thrown if the currently checked out GIT branch is null or empty..</exception>
```
Comment (kselena) [inline CICD/Services/BranchValidatorService.cs:617]: remove extra period
```suggestion
    /// <exception cref="Exception">Thrown if the currently checked out GIT branch is null or empty.</exception>
```

File: CICD/Services/BranchValidatorService.cs [hunk 6]
```diff
@@ -0,0 +1,700 @@
...
+        if (branch.IsHotFixBranch())
+        {
+            return BranchType.HotFix;
+        }
+
+        return BranchType.Other;
+    }
+
+    /// <inheritdoc/>
+    /// <exception cref="Exception">Thrown if the currently checked out GIT branch is null or empty..</exception>
```
Comment (kselena) [inline CICD/Services/BranchValidatorService.cs:603]: remove extra period
```suggestion
    /// <exception cref="Exception">Thrown if the currently checked out GIT branch is null or empty.</exception>
```

File: CICD/Services/IBranchValidatorService.cs [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+    /// <summary>
+    /// Validates that the given <paramref name="branch"/> has the correct syntax using the given <paramref name="branchPattern"/> and
+    /// combines the given <paramref name="andPredicates"/> results with the function result.
+    /// </summary>
+    /// <param name="branch">The GIT branch.</param>
+    /// <param name="branchPattern">The syntax pattern.</param>
+    /// <param name="andPredicates">The list of predicates to include in the end result.</param>
+    /// <returns>The current object to continue the <c>Fluent API</c> chain.</returns>
+    /// <remarks>
+    ///     If any of the <paramref name="andPredicates"/> return false, the entire function will return false.
```
Comment (kselena) [inline CICD/Services/IBranchValidatorService.cs:31]:
```suggestion
    /// If any of the <paramref name="andPredicates"/> return false, the entire function will return false.
```

File: CICD/Services/IBranchValidatorService.cs [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+    /// <summary>
+    /// Validates that the given <paramref name="branch"/> matches the given <paramref name="branchType"/> and
+    /// combines the given <paramref name="andPredicates"/> results with the function result.
+    /// </summary>
+    /// <param name="branch">The GIT branch.</param>
+    /// <param name="branchType">The type of branch to expect.</param>
+    /// <param name="andPredicates">The list of predicates to include in the end result.</param>
+    /// <returns>The current object to continue the <c>Fluent API</c> chain.</returns>
+    /// <remarks>
+    ///     If any of the <paramref name="andPredicates"/> return false, the entire function will return false.
```
Comment (kselena) [inline CICD/Services/IBranchValidatorService.cs:52]:
```suggestion
    /// If any of the <paramref name="andPredicates"/> return false, the entire function will return false.
```

File: CICD/Services/IBranchValidatorService.cs [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+
+    /// <summary>
+    /// Validates if the given <paramref name="branch"/> is a <see cref="BranchType.Feature"/> branch and
+    /// combines the given <paramref name="andPredicates"/> results with the function result.
+    /// </summary>
+    /// <param name="branch">The GIT branch.</param>
+    /// <param name="andPredicates">The list of predicates to include in the end result.</param>
+    /// <returns>The current object to continue the <c>Fluent API</c> chain.</returns>
+    /// <remarks>
+    ///     If any of the <paramref name="andPredicates"/> return false, the entire function will return false.
```
Comment (kselena) [inline CICD/Services/IBranchValidatorService.cs:109]:
```suggestion
    /// If any of the <paramref name="andPredicates"/> return false, the entire function will return false.
```

File: CICD/Services/IBranchValidatorService.cs [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+
+    /// <summary>
+    /// Validates that the given <paramref name="branch"/> is a <see cref="BranchType.PreviewFeature"/> branch and
+    /// combines the given <paramref name="andPredicates"/> results with the function result.
+    /// </summary>
+    /// <param name="branch">The GIT branch.</param>
+    /// <param name="andPredicates">The list of predicates to include in the end result.</param>
+    /// <returns>The current object to continue the <c>Fluent API</c> chain.</returns>
+    /// <remarks>
+    ///     If any of the <paramref name="andPredicates"/> return false, the entire function will return false.
```
Comment (kselena) [inline CICD/Services/IBranchValidatorService.cs:147]:
```suggestion
    /// If any of the <paramref name="andPredicates"/> return false, the entire function will return false.
```

File: CICD/Services/IBranchValidatorService.cs [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+
+    /// <summary>
+    /// Validates that the given <paramref name="branch"/> is a <see cref="BranchType.Release"/> branch and
+    /// combines the given <paramref name="andPredicates"/> results with the function result.
+    /// </summary>
+    /// <param name="branch">The GIT branch.</param>
+    /// <param name="andPredicates">The list of predicates to include in the end result.</param>
+    /// <returns>The current object to continue the <c>Fluent API</c> chain.</returns>
+    /// <remarks>
+    ///     If any of the <paramref name="andPredicates"/> return false, the entire function will return false.
```
Comment (kselena) [inline CICD/Services/IBranchValidatorService.cs:166]:
```suggestion
    /// If any of the <paramref name="andPredicates"/> return false, the entire function will return false.
```

File: CICD/Services/IBranchValidatorService.cs [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+
+    /// <summary>
+    /// Validates that the given <paramref name="branch"/> is a <see cref="BranchType.HotFix"/> branch and
+    /// combines the given <paramref name="andPredicates"/> results with the function result.
+    /// </summary>
+    /// <param name="branch">The GIT branch.</param>
+    /// <param name="andPredicates">The list of predicates to include in the end result.</param>
+    /// <returns>The current object to continue the <c>Fluent API</c> chain.</returns>
+    /// <remarks>
+    ///     If any of the <paramref name="andPredicates"/> return false, the entire function will return false.
```
Comment (kselena) [inline CICD/Services/IBranchValidatorService.cs:185]:
```suggestion
    /// If any of the <paramref name="andPredicates"/> return false, the entire function will return false.
```

File: CICD/Services/IBranchValidatorService.cs [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+    /// <summary>
+    /// Validates that the target branch of a pull request that matches the given <paramref name="prNumber"/> matches
+    /// the given <paramref name="branchType"/> and combines the given <paramref name="andPredicates"/> results with the function result.
+    /// </summary>
+    /// <param name="prNumber">The pull request number.</param>
+    /// <param name="branchType">The type of branch to expect.</param>
+    /// <param name="andPredicates">The list of predicates to include in the end result.</param>
+    /// <returns>The current object to continue the <c>Fluent API</c> chain.</returns>
+    /// <remarks>
+    ///     If any of the <paramref name="andPredicates"/> return false, the entire function will return false.
```
Comment (kselena) [inline CICD/Services/IBranchValidatorService.cs:229]:
```suggestion
    /// If any of the <paramref name="andPredicates"/> return false, the entire function will return false.
```

File: CICD/Services/IBranchValidatorService.cs [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+
+    /// <summary>
+    /// Provides the ability to use <c>or</c> logic with the results between other <c>Fluent API</c> methods.
+    /// </summary>
+    /// <returns>The current object to continue the <c>Fluent API</c> chain.</returns>
+    IBranchValidatorService Or();
+
+    /// <summary>
+    /// Returns a <c>bool</c> value indicating whether or not the given <paramref name="branch"/> that <c>might</c> contain
+    /// and issue number exists.
```
Comment (kselena) [inline CICD/Services/IBranchValidatorService.cs:241]: grammar
```suggestion
    /// Returns a <c>bool</c> value indicating whether or not the given <paramref name="branch"/> <c>might</c> contain
    /// an existing issue number.
```

File: CICD/Services/IBranchValidatorService.cs [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+    /// <param name="branchPattern">The syntax pattern.</param>
+    /// <returns><c>true</c> if the a match.</returns>
+    bool CurrentBranchIsValid(string branchPattern);
+
+    /// <summary>
+    /// Returns a <c>bool</c> value indicating whether or not the currently checked out GIT branch
+    /// matches the given <paramref name="branchType"/>.
+    /// </summary>
+    /// <param name="branchType">The type of branch to expect.</param>
+    /// <returns><c>true</c> if the a match.</returns>
```
Comment (kselena) [inline CICD/Services/IBranchValidatorService.cs:267]: grammar
```suggestion
    /// <returns><c>true</c> if there is a match.</returns>
```

File: CICD/Services/IBranchValidatorService.cs [hunk 16]
```diff
@@ -0,0 +1,281 @@
...
+    bool CurrentBranchIsValid(BranchType branchType);
+
+    /// <summary>
+    /// Resets the internal state of the objects logic.
```
Comment (kselena) [inline CICD/Services/IBranchValidatorService.cs:271]: grammar
```suggestion
    /// Resets the internal state of the object's logic.
```

File: CICD/Services/IPullRequestService.cs [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+    /// <summary>
+    /// Gets the source branch of a pull request.
+    /// </summary>
+    /// <remarks>This is the branch that is being merged into the <see cref="TargetBranch"/>.</remarks>
+    string SourceBranch { get; }
+
+    /// <summary>
+    /// Gets the target branch of a pull request.
+    /// </summary>
+    /// <remarks>This is the branch that a pull request <see cref="SourceBranch"/> merging into.</remarks>
```
Comment (kselena) [inline CICD/Services/IPullRequestService.cs:21]: grammar
```suggestion
    /// <remarks>This is the branch where a pull request <see cref="SourceBranch"/> merges.</remarks>
```

File: CICD/Services/IPullRequestService.cs [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+    string TargetBranch { get; }
+
+    /// <summary>
+    /// Gets the pull request number of the current build.
+    /// </summary>
+    /// <returns>The pull request number.</returns>
+    /// <remarks>
+    ///     Returns the value -1 if no pull request number exists.
```
Comment (kselena) [inline CICD/Services/IPullRequestService.cs:29]:
```suggestion
    /// Returns the value -1 if no pull request number exists.
```

**Review by kselena — CHANGES_REQUESTED** [review]

File: CICD/Reactables/Core/IReactor.cs [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+    /// Provides the reactor with new data.
+    /// </summary>
+    /// <param name="value">The current notification information.</param>
+    public void OnNext(T value);
+
+    /// <summary>
+    /// Notifies the reactor that the provider has finished sending push-based notifications.
+    /// </summary>
+    /// <remarks>
+    ///     Will not be invoked more than once.
```
[the remark this answers is not in the document]
Comment (CalvinWilkinson) [reply CICD/Reactables/Core/IReactor.cs:25]: It is ok to have some XML content in the code docs be indented if the starting element and ending element are on separate lines.  This increases readability.

**Review by kselena — DISMISSED** [review]

**Review by KinsonDigitalAdmin — APPROVED** [review]

**Review by kselena — APPROVED** [review]
